test/test_optim.py

```diff
@@ -294,47 +295,48 @@ def test_sgd_sparse(self):
         )
 
     def test_adam(self):
-        self._test_basic_cases(
-            lambda weight, bias: optim.Adam([weight, bias], lr=1e-3)
-        )
-        self._test_basic_cases(
-            lambda weight, bias: optim.Adam(
-                self._build_params_dict(weight, bias, lr=1e-2),
-                lr=1e-3)
-        )
-        self._test_basic_cases(
-            lambda weight, bias: optim.Adam([weight, bias], lr=1e-3,
-                                            amsgrad=True)
-        )
-        self._test_basic_cases(
-            lambda weight, bias: optim.Adam(
-                self._build_params_dict(weight, bias, lr=1e-2),
-                lr=1e-3, amsgrad=True)
-        )
-        self._test_basic_cases(
-            lambda weight, bias: optim.Adam(
-                self._build_params_dict(weight, bias, lr=1e-2),
-                lr=1e-3),
-            [lambda opt: ExponentialLR(opt, gamma=0.9)]
-        )
-        self._test_basic_cases(
-            lambda weight, bias: optim.Adam([weight, bias], lr=1e-3,
-                                            amsgrad=True),
-            [lambda opt: ExponentialLR(opt, gamma=0.9),
-             lambda opt: ReduceLROnPlateau(opt)]
-        )
-        self._test_basic_cases(
-            lambda weight, bias: optim.Adam(
-                self._build_params_dict(weight, bias, lr=1e-2),
-                lr=1e-3, amsgrad=True),
-            [lambda opt: StepLR(opt, gamma=0.9, step_size=10),
-             lambda opt: ReduceLROnPlateau(opt)]
-        )
-        with self.assertRaisesRegex(ValueError, "Invalid beta parameter at index 0: 1.0"):
-            optim.Adam(None, lr=1e-2, betas=(1.0, 0.0))
-
-        with self.assertRaisesRegex(ValueError, "Invalid weight_decay value: -1"):
-            optim.Adam(None, lr=1e-2, weight_decay=-1)
+        for optimizer in [optim.Adam, optim_mt.Adam]:
```

Comment:
note: current tests don't guarantee that the algorithm is the same as the other, or even that there's convergence in either case.

@zou3519 @anjali411 -- how are the C++ APIs tested?

Reply:
C++ Optimizer API logic is tested here: https://github.com/pytorch/pytorch/blob/master/test/cpp/api/optim.cpp#L311. These tests compare the C++ optimizers' results to the Python API optimizers' results prewritten in this file: https://github.com/pytorch/pytorch/blob/master/test/cpp/api/optim_baseline.h

Reply:
Ok, so here we'll need the tests comparing optim.Adam results with optim_mt.Adam, in addition to `_test_basic_cases`?

Reply:
Yes, since the contract here is that the MultiTensor implementation is the same as the original.

Reply:
Is there also a C++ equivalent with the foreach API for the C++ optimizer?

Reply:
@vincentqb, there will be C++ optimizers as well but a bit later. We decided to start with python ones first. 

re testing: Is there anything specific you would suggest testing?

Reply:
Ideally, we'd have a test as mentioned by @anjali411 [above](https://github.com/pytorch/pytorch/pull/43507#discussion_r486502986) that checks that the two implementations give the exact same answer in some case.

Note that if this implementation were directly replacing the current one, the C++ test would also tell us that the implementations are still aligned :) Maybe there's a way of leveraging those tests already there?

Reply:
There are tests in apex comparing optimizer implementations, they can be adopted here if c++ tests are hard to use for some reason https://github.com/NVIDIA/apex/blob/master/tests/L0/run_optimizers/test_fused_optimizer.py